---
title: "Bacterial Genome Analysis Report"
author: "Harshada Pardeshi"
date: "`r Sys.Date()`"
format:
  html: default
execute: 
  freeze: false
editor: visual
---

```{r libraries, include=FALSE, echo=FALSE, warning=FALSE}
options(repos = c(CRAN = "https://cran.rstudio.com/"))
required_packages <- c("BiocManager", "knitr", "XVector","Biostrings", "kableExtra", "dplyr", "rmarkdown", "rvest", "jsonlite", "readr", "data.table", "xml2","devtools","fastqcr","textshaping")

for (pkg in required_packages) {
  if (!requireNamespace(pkg, quietly = TRUE)) {
    install.packages(pkg)
  }
  library(pkg, character.only = TRUE)
}

file_path <- "absolute_paths.txt"
```

## Assembly

::: panel-tabset
### Assembly

After quality check with FastQC, raw reads were trimmed based on their quality and length using [Trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic). The default criteria for trimming was `SLIDINGWINDOW:4:15`and `MINLEN:36`. The option `SLIDINGWINDOW:4:15` means that the read was scanned with a sliding window of size 4 bases. If the average quality within the window dropped below 15, the read was trimmed from that point onwards. `MINLEN:36` ensures that only reads longer than or equal to 36 bases are kept after trimming. Reads shorter than 36 bases are discarded.

Trimmed reads were used for assembling the genome using [Unicycler](https://github.com/rrwick/Unicycler).

Unicycler is an assembly pipeline for bacterial genomes. It can assemble Illumina-only read sets where it functions as a SPAdes-optimiser. It produces assemblies that are accurate, complete and cost-effective. Unicycler builds an initial assembly graph from short reads using the de novo assembler SPAdes and then simplifies the graph using information from short and long reads.It can also assemble long-read-only sets and hybrid data sets.

Trimmed reads are stored in `00_trimmed_reads` and Unicycler assembly output files are stored in `02_unicycler` directory in the specified `outdir`.

Following is the command used for assembling the genome:

``` bash
unicycler -1 {sample_dir/00_trimmed_reads/forward_paired.fastq.gz} -2 {sample_dir/00_trimmed_reads/reverse_paired.fastq.gz} -o {sample_dir/02_unicycler} --min_fasta_length 1000 -t 40 &> {sample_dir/logs/unicycler.log}
```

Given below is the information of contigs generated by Unicycler assembly

```{r unicycler, echo=FALSE}
paths <- readLines(file_path)
unicycler <- paths[3]

# Read the multifasta file
fasta_file <- readDNAStringSet(unicycler)
#| echo: false
#|
# Extract information from headers
contig_info <- names(fasta_file)

# Function to extract values using regular expressions
extract_info <- function(header) {
  contig_num <- as.numeric(sub(">", "", sub(" .*", "", header)))
  length <- as.numeric(sub(".*length=", "", sub(" depth=.*", "", header)))
  depth <- as.numeric(sub(".*depth=", "", sub("x", "", header)))
  return(c(contig_num, length, depth))
}

# Apply the function to each header
extracted_data <- t(sapply(contig_info, extract_info))

# Create a data frame
contig_df <- data.frame(
  Contig = extracted_data[, 1],
  Length = extracted_data[, 2],
  Depth = extracted_data[, 3]
)

# Print the total number of contigs in bold
#cat("**Total number of contigs generated after assembly are:**", length(contig_df$Contig), "\n")
knitr::asis_output(paste("**Total number of contigs generated after assembly are:", length(contig_df$Contig), "**\n"))

# Display the table using kable() and kableExtra
kable(contig_df, caption = "Unicycler Assembly Contig Information") %>%
  kable_styling(bootstrap_options = "striped", full_width = FALSE, position = "center") %>%
row_spec(0, bold = TRUE) %>% # Bold the header row
scroll_box(width = "900px", height = "600px")

# Summarize the data
contig_summary <- contig_df %>%
  summarize(
    Mean_Length = mean(Length),
    Median_Length = median(Length),
    Max_Length = max(Length),
    Min_Length = min(Length),
    )

# Display the summary table using kable() and kableExtra
kable(contig_summary, results = "asis", align = "c", caption = "Contig Summary") %>%
  kable_styling(full_width = FALSE, position = "center", bootstrap_options = "striped") %>%
row_spec(0, bold = TRUE) %>% # Bold the header row
scroll_box(width = "900px", height = "300px")


```

For more details, Unicycler contig fasta file is saved here

| Filename                                                | Location                | Description                     |
|------------------------------|-------------------|-----------------------|
| [assembly.fasta](file://`r file.path(unicycler)`) | sample_dir/02_unicycler | Fasta file of assembled contigs |

### Scaffolding

After performing assembly with Unicycler, the assembled contigs were ordered and oriented into longer sequences using [RagTag](https://github.com/malonge/RagTag).

Scaffolds are stored in `06_ragtag` directory in the specified `outdir`.

Following is the command used for scaffolding the genome:

``` bash
ragtag.py scaffold {reference.fasta} {sample_dir/02_unicycler/assembly.fasta} -t 20 -o {sample_dir/06_ragtag} &> {sample_dir/logs/ragtag.log}
```

```{r ragtag, echo=FALSE}
paths <- readLines(file_path)
ragtag <- paths[12]
fasta_file <- readDNAStringSet(ragtag)
num_sequences <- length(fasta_file)
#cat("**Number of sequences after scaffolding is", num_sequences, "**\n")
knitr::asis_output(paste("**Total number of scaffolds generated after scaffolding are:", num_sequences, "**\n"))
```

For more details, RagTag scaffolds fasta file is saved here

| Filename                                                     | Location             | Description             |
|------------------------------|-------------------|-----------------------|
| [ragtag.scaffolds.fasta](file://`r file.path(ragtag)`) | sample_dir/06_ragtag | Fasta file of scaffolds |

### Polishing

After performing scaffolding with RagTag, [Pilon](https://github.com/broadinstitute/pilon) was used for improving the draft assembly. Pilon requires as input a FASTA file of the genome along with one or more BAM files of reads aligned to the input FASTA file. It uses read alignment analysis to identify inconsistencies between the input genome and the evidence in the reads. It then attempts to make improvements to the input genome, including single base differences, small indels, larger indel or block substitution events, gap filling, identification of misassemblies.

[BWA](https://github.com/lh3/bwa) was used to generate a BAM file by mapping raw reads against scaffolds generated by RagTag. BWA-MEM algorithm was used as it is faster and more accurate.

BAM file is stored in `07_bwa` directory in the specified `outdir`.

Pilon output FASTA file (polished genome) is used for functional annotation.

Polished genome is stored in `08_pilon` directory in the specified `outdir`.

Following is the command used for assembling the genome:

``` bash
pilon --threads 40 --genome {sample_dir/06_ragtag/ragtag.scaffold.fasta} --frags {sample_dir/07_bwa/{sample_name}.bam} --outdir {sample_dir/08_pilon} --output {sample_name} -Xmx 160G &> {sample_dir/logs/pilon.log}
sed 's/|/_/g' {sample_dir/08_pilon/{sample_name}.fasta} > {sample_dir/08_pilon/{sample_name}_contig_header_reduced.fasta}
count=0
        while IFS= read -r line
        do
            if [[ $line == ">"* ]]; then
                count=$((count + 1))
                echo ">contig_$count" >> {sample_dir/08_pilon/temp.fasta}
            else
                echo "$line" >> {sample_dir/08_pilon/temp.fasta}
            fi
        done < {sample_dir/08_pilon/{sample_name}_contig_header_reduced.fasta}
        rm {sample_dir/08_pilon/{sample_name}_contig_header_reduced.fasta}
        mv {sample_dir/08_pilon/temp.fasta} {sample_dir/08_pilon/{sample_name}_contig_header_reduced.fasta}
```

```{r pilon, echo=FALSE}
paths <- readLines(file_path)
pilon <- paths[13]
```

For more details, polished genome fasta file is saved here

| Filename                                         | Location            | Description                   |
|------------------------------|-------------------|-----------------------|
| [pilon.fasta](file://`r file.path(pilon)`) | sample_dir/08_pilon | Fasta file of polished genome |
:::
