---
title: "Bacterial Genome Analysis Report"
author: "Harshada Pardeshi"
date: "`r Sys.Date()`"
format:
  html: default
execute: 
  freeze: false
editor: visual
---

```{r libraries, include=FALSE, echo=FALSE, warning=FALSE}
options(repos = c(CRAN = "https://cran.rstudio.com/"))
required_packages <- c("BiocManager", "knitr", "XVector","Biostrings", "kableExtra", "dplyr", "rmarkdown", "rvest", "jsonlite", "readr", "data.table", "xml2","devtools","fastqcr","textshaping")

for (pkg in required_packages) {
  if (!requireNamespace(pkg, quietly = TRUE)) {
    install.packages(pkg)
  }
  library(pkg, character.only = TRUE)
}

file_path <- "absolute_paths.txt"
```

# Assembly

::: panel-tabset
## Assembly

After performing quality check with Chopper, the filtered reads were used for de novo genome assembly using [Flye](https://github.com/mikolmogorov/Flye).

Flye is a de novo assembler for single-molecule sequencing reads. It represents a complete pipeline: it takes raw PacBio / ONT reads as input and outputs polished contigs. It doesn't perform scaffolding by default, so it is enabled by using `--scaffold` option in the flye command as mentioned below.

Flye assembly FASTA file and other output files is stored in `01_flye` in the specified `outdir`.

Following is the command used for assembling the genome with ONT reads:

``` bash
mkdir -p 01_flye
flye --nano-corr/--nano-raw/--nano-hq {sample_dir/00_chopper/filtered.fastq} --out-dir {sample_dir/01_flye} --scaffold 2> {sample_dir/logs/flye.log}
```

Following is the command used for assembling the genome with PacBio reads:

``` bash
mkdir -p 01_flye
flye --pacbio-corr/--pacbio-raw/--pacbio-hifi {sample_dir/00_chopper/filtered.fastq} --out-dir {sample_dir/01_flye} --scaffold 2> {sample_dir/logs/flye.log}
```

Given below is the information of genome assembly

```{r flye, echo=FALSE}
paths <- readLines(file_path)
flye <- paths[3]
fasta_file <- readDNAStringSet(flye)
num_sequences <- length(fasta_file)
cat("Number of sequences generated by Flye assembly are", num_sequences, "\n")
```

```{r flye_stats, echo=FALSE}
paths <- readLines(file_path)
flye_stats <- paths[43]
file_content <- readLines(flye_stats)
knitr::asis_output("Contigs summary statistics:")
knitr::asis_output(paste(file_content, collapse = "\n"))
```

For more details, flye contig fasta file is saved here

| Filename                                                    | Location           | Description                         |
|------------------------------|-------------------|-----------------------|
| [assembly.fasta](file://`r file.path(flye)`)          | sample_dir/02_flye | Fasta file of assembled contigs     |
| [assembly_info.txt](file://`r file.path(flye_stats)`) | sample_dir/02_flye | Summary file of contigs information |

## Polishing

After performing genome assembly with Flye, [Pilon](https://github.com/broadinstitute/pilon) was used for improving the draft assembly. Pilon requires as input a FASTA file of the genome along with one or more BAM files of reads aligned to the input FASTA file. It uses read alignment analysis to identify inconsistencies between the input genome and the evidence in the reads. It then attempts to make improvements to the input genome, including single base differences, small indels, larger indel or block substitution events, gap filling, identification of misassemblies. 

[BWA](https://github.com/lh3/bwa) was used to generate a BAM file by mapping raw reads against scaffolds generated by RagTag. BWA-MEM algorithm was used as it is faster and more accurate.

BAM file is stored in `07_bwa` directory in the specified `outdir`.

Pilon output FASTA file (polished genome) is used for functional annotation.

Polished genome is stored in `08_pilon` directory in the specified `outdir`.

Following is the command used for assembling the genome:

``` bash
pilon --threads 40 --genome {sample_dir/06_ragtag/ragtag.scaffold.fasta} --frags {sample_dir/07_bwa/{sample_name}.bam} --outdir {sample_dir/08_pilon} --output {sample_name} -Xmx 160G &> {sample_dir/logs/pilon.log}
sed 's/|/_/g' {sample_dir/08_pilon/{sample_name}.fasta} > {sample_dir/08_pilon/{sample_name}_contig_header_reduced.fasta}
count=0
        while IFS= read -r line
        do
            if [[ $line == ">"* ]]; then
                count=$((count + 1))
                echo ">contig_$count" >> {sample_dir/08_pilon/temp.fasta}
            else
                echo "$line" >> {sample_dir/08_pilon/temp.fasta}
            fi
        done < {sample_dir/08_pilon/{sample_name}_contig_header_reduced.fasta}
        rm {sample_dir/08_pilon/{sample_name}_contig_header_reduced.fasta}
        mv {sample_dir/08_pilon/temp.fasta} {sample_dir/08_pilon/{sample_name}_contig_header_reduced.fasta}
```
```{r pilon, echo=FALSE}
paths <- readLines(file_path)
pilon <- paths[4]
fasta_file <- readDNAStringSet(pilon)
num_sequences <- length(fasta_file)
cat("Number of sequences generated after polishing the assembly are", num_sequences, "\n")
```
For more details, polished genome fasta file is saved here

| Filename                                                   | Location                | Description                     |
|------------------------------|-------------------|-----------------------|
| [pilon.fasta](file://`r file.path(pilon)`) | sample_dir/04_pilon | Fasta file of polished genome |
:::