---
title: "Bacterial Genome Analysis Report"
author: "Harshada Pardeshi"
date: "`r Sys.Date()`"
format:
  html: default
execute: 
  freeze: false
editor: visual
---

```{r libraries, include=FALSE, echo=FALSE, warning=FALSE}
options(repos = c(CRAN = "https://cran.rstudio.com/"))
required_packages <- c("BiocManager", "knitr", "XVector","Biostrings", "kableExtra", "dplyr", "rmarkdown", "rvest", "jsonlite", "readr", "data.table", "xml2","devtools","fastqcr","textshaping")

for (pkg in required_packages) {
  if (!requireNamespace(pkg, quietly = TRUE)) {
    install.packages(pkg)
  }
  library(pkg, character.only = TRUE)
}

file_path <- "absolute_paths.txt"
```

# Assembly

::: panel-tabset
## Assembly

After performing quality check with Chopper, the filtered reads were used for de novo genome assembly using [Flye](https://github.com/mikolmogorov/Flye).

Flye is a de novo assembler for single-molecule sequencing reads. It represents a complete pipeline: it takes raw PacBio / ONT reads as input and outputs polished contigs. It doesn't perform scaffolding by default, so it is enabled by using `--scaffold` option in the flye command as mentioned below.

Flye assembly FASTA file and other output files is stored in `01_flye` in the specified `outdir`.

Following is the command used for assembling the genome with ONT reads:

``` bash
mkdir -p 01_flye
flye --nano-corr/--nano-raw/--nano-hq {sample_dir/00_chopper/filtered.fastq} --out-dir {sample_dir/01_flye} --scaffold 2> {sample_dir/logs/flye.log}
```

Following is the command used for assembling the genome with PacBio reads:

``` bash
mkdir -p 01_flye
flye --pacbio-corr/--pacbio-raw/--pacbio-hifi {sample_dir/00_chopper/filtered.fastq} --out-dir {sample_dir/01_flye} --scaffold 2> {sample_dir/logs/flye.log}
```

Given below is the information of genome assembly

```{r flye, echo=FALSE}
paths <- readLines(file_path)
flye <- paths[1]
fasta_file <- readDNAStringSet(flye)
num_sequences <- length(fasta_file)
#cat("Number of sequences generated by Flye assembly are", num_sequences, "\n")
knitr::asis_output(paste("**Total number of scaffolds generated after scaffolding are:", num_sequences, "**\n"))
```

```{r flye_stats, echo=FALSE}
library(knitr)
library(kableExtra)
paths <- readLines(file_path)
flye_stats <- paths[41]
file_content <- read.table(flye_stats, header = TRUE, sep = "\t", fill = TRUE)

kable(file_content, caption = "Contigs Summary Statistics") %>%
  kable_styling(bootstrap_options = "striped", full_width = FALSE, position = "center") %>%
row_spec(0, bold = TRUE) %>% # Bold the header row
scroll_box(width = "900px", height = "600px")
```

For more details, flye contig fasta file is saved here

| Filename                                                    | Location           | Description                         |
|------------------------------|-------------------|-----------------------|
| [assembly.fasta](file://`r file.path(flye)`)          | sample_dir/02_flye | Fasta file of assembled contigs     |
| [assembly_info.txt](file://`r file.path(flye_stats)`) | sample_dir/01_flye | Summary file of contigs information |

## Polishing

After performing genome assembly with Flye, [Racon](https://github.com/isovic/racon) was used for improving the draft assembly. The goal of Racon is to generate genomic consensus which is of similar or better quality compared to the output generated by assembly methods which employ both error correction and consensus steps, while providing a speedup of several times compared to those methods. It supports data produced by both PacBio and ONT.

Along with the assembled FASTA file and raw reads FASTQ file, Racon also requires a alignment SAM file which is generated using [Minimap2](https://github.com/lh3/minimap2).

SAM file is stored in `02_minimap2` directory in the specified `outdir`.

Racon output FASTA file is used for functional annotation.

Polished genome is stored in `03_racon` directory in the specified `outdir`.

Following is the command used for assembling the genome:

``` bash
        racon {raw_reads.fastq} {sample_dir/02_minimap2/aln.sam} {sample_dir/01_flye/assembly.fasta} > {sample_dir/03_racon/polish.fasta} 2> {sample_dir/logs/racon.log}
        sed 's/|/_/g' {sample_dir/03_racon/polish.fasta} > {sample_dir/03_racon/{sample}_final.fasta}
        count=0
        while IFS= read -r line
        do
            if [[ $line == ">"* ]]; then
                count=$((count + 1))
                echo ">contig_$count" >> {sample_dir/03_racon/temp.fasta}
            else
                echo "$line" >> {sample_dir/03_racon/temp.fasta}
            fi
        done < {sample_dir/03_racon/{sample}_final.fasta}
        rm {sample_dir/03_racon/{sample}_final.fasta}
        mv {sample_dir/03_racon/temp.fasta} {sample_dir/03_racon/{sample}_final.fasta}
```

```{r racon, echo=FALSE}
paths <- readLines(file_path)
racon <- paths[2]
```

For more details, polished genome fasta file is saved here

| Filename                                         | Location            | Description                   |
|------------------------------|-------------------|-----------------------|
| [pilon.fasta](file://`r file.path(racon)`) | sample_dir/03_racon | Fasta file of polished genome |
:::
